### ISSUE TITLE
Support for Node ^v12

### ISSUE THREAD

**Issue by Amirh24** [aws/aws-xray-sdk-node#154]
Do you plan to support later versions of node in the future (^v10)?
We are using your sdk but it limits us to node < v10 and under which makes us not able to use libraries that are only compatible with the recent versions.

Thanks

**Comment by awssandra** [comment]
Hi Amirh24,

We [support Node 12.x](https://github.com/aws/aws-xray-sdk-node#installing) and will continue to support newer versions.

**Comment by Amirh24** [comment]
Hi Sandra, thanks for your replay.

Per you github page:
"The AWS X-Ray SDK for Node.js is compatible with Node.js version 4 and later. The AWS X-Ray SDK for Node.js has been tested with versions 4.x through 12.x of Node.js. There may be issues when running on versions of Node.js newer than 12.x."

Would you say that it is safe to use the AWS X-RAY SDK that is compatible with Node12.x on an app in production?

Thanks

**Comment by awssandra** [comment]
Yes, there should be no issues with Node12. The comment "There may be issues when running on versions of Node.js newer than 12.x" merely means, as new Node releases come out, we do not guarantee compatibility from day 1, or with pre-release versions, as there may be breaking changes in compatibility. As a part of our continuous integration tests, we have them configured to automatically run per each commit on the repository for node versions up to 12. Thus, we test internally and upgrade our CI tests to cover the new version, before we announce support.

Hope this helps,
Sandra

**Comment by wparad** [comment]
As far as I can tell, this isn't an appropriate response to the question. Let me know if I'm mistaken. We suggest that we want to support node 12, which likely means that the client code is using `async/await`. 

Does this library support async code being executed inside the `ns.run` block?

**Comment by awssandra** [comment]
Hi wparad,

Can you clarify what you mean? As for async/await structures, those are supported via the experimental branch, which we will be integrating into our 3.0 major version as per our [pinned issue](https://github.com/aws/aws-xray-sdk-node/issues/157). More info on [our experimental branch](https://github.com/aws/aws-xray-sdk-node/issues/60).

As per the above discussions, the decision to keep our branch experimental was under the fact the [Node async-hooks APIs themselves](https://nodejs.org/api/async_hooks.html#async_hooks_public_api) used the async/await-compatible CLS library are [marked as experimental](https://nodejs.org/api/documentation.html#documentation_stability_index), and have yet to be reassessed by the Node team. However, [many GA'd open-source libraries are already using async-hooks as a part of their context propagation solutions used in production environments](https://github.com/aws/aws-xray-sdk-node/issues/60#issuecomment-496333896) thus we are comfortable moving this into our next major release.

The async function handler on Lambda runtimes is a separate issue, which is being followed up on by the AWS Lambda team.

Hope this helps,
Sandra

**Comment by wparad** [comment]
It wasn't clear what needed to be done by a developer to actually get this working. Perhaps it would help if some light was cast on the following issues. I found it difficult to get them to work:

* Directly adding annotations to the current lambda segment, does not show in the xray console.
* `AWSXray.getSegment().addNewSubsegment('new-segment')` does not show up in the xray console attached to the Lambda::Function or anywhere I could find.
* When `AWSXray.captureAsyncFunc()` is used, it doesn't wrap the async calls to S3 or dynamoDB.
* XRay segments aren't sent until the next iteration of the lambda function, for some reason blocking on the `await AWSXray.captureAsyncFunc('type', async () => {})` doesn't actually wait on the `SegmentEmitter` to be completed.

What I would like to do inside lambda (without express) is to:
* create a new subsegment of the existing lambda execution
* have all the further XRay segments automatically created by the XRay sdk wrapping http or aws calls to appear inside of this new segment.

What I'm confused with is that there are no docs for how to do with out using express. Assuming I just have a lambda and no framework, can a code example be provided?

**Comment by willarmiros** [comment]
Feel free to reopen if you have further questions about the usage of the SDK.
